include/hipSYCL/sycl/libkernel/builtin_interface.hpp

```diff
@@ -181,8 +181,8 @@ HIPSYCL_BUILTIN T __acpp_fmod(T x, T y) noexcept {
   HIPSYCL_RETURN_DISPATCH_BUILTIN(__acpp_fmod, x, y);
 }
 
-template<class T>
-T __acpp_fract(T x, T* ptr) noexcept {
+template<class T, class FloatPtr>
+T __acpp_fract(T x, FloatPtr ptr) noexcept {
```

Comment:
I think this will fail for `multi_ptr` or other pointer types, as the pointer here is directly forwarded to the dispatch-builtin macro and may therefore end up e.g. in SSCP builtin calls which only exist for specific raw pointer types.

If we allow any type that provides a pointer interface, we should ensure at a high-level that the type that is actually passed into backend builtins remains a supported pointer type.

Reply:
Yeah, my ideal solution would be to change the `HIPSYCL_RETURN_DISPATCH_BUILTIN` call to handle passing it in a raw pointer, as that would be a decent decrease in complexity, but the functions here directly delegate and I wasn't sure if that was intentional. Something like the following would be ideal imo (and could provide future hooks for private addr spaces and all:

```
template<class T, class FloatPtr, std::enable_if_t<std::is_same_v<FloatPtr, T*>, bool> = true>
T __acpp_fract(T x, FloatPtr ptr) noexcept {
  HIPSYCL_RETURN_DISPATCH_BUILTIN(__acpp_fract, x, ptr);
}

template<class T, class FloatPtr, std::enable_if_t<!std::is_same_v<FloatPtr, T*>, bool> = true>
T __acpp_fract(T x, FloatPtr ptr) noexcept {
  HIPSYCL_RETURN_DISPATCH_BUILTIN(__acpp_fract, x, ptr.get());
}
```

Reply:
It seems that I was mistaken by how some of the builtins implementation worked, and it appears that this wouldn't really solve the problem. I've ripped out the changes related to this until I can come back and take a better look at the whole dispatch chain.